### PR TITLE
(PC-4934) New password : Handle deeplink and redirection to the right screen

### PR DESCRIPTION
**Pull request by blixit**
Link to JIRA ticket: https://passculture.atlassian.net/browse/PC-XXX

## Checklist

I have:

- [x] Made sure the title of my PR follows the convention `[PC-XXX] <summary>`.
- [ ] Made sure my feature is working on the relevant real / virtual devices.
- [x] Written **unit tests** for my feature.
- [ ] Added a **screenshot** for UI tickets.
- [ ] Explained my technical strategy below if feature is complex.

image

image

## Deploy hard

If native code (ios/android) was modified, **after** the PR is merged, on the master branch, upgrade the **app version** (+1 patch):

- if you want an hard deployment of the testing environment, use `yarn version:testing` (this will create a commit with a tag)
- then run `git push --follow-tags`

## Technical strategy

> _Insert technical strategy_

### CODE REVIEW

File: src/features/deeplinks/useDeeplinkUrlHandler.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  return `${env.URL_PREFIX}://${env.IOS_APP_ID}/`
+}
+
+export function formatAndroidDeeplinkDomain() {
+  return `${env.URL_PREFIX}://${env.ANDROID_APP_ID}/`
+}
+
+export function decodeDeeplinkParts(url: string): DeeplinkParts {
+  let route = url.replace(formatIosDeeplinkDomain(), '')
+  route = route.replace(formatAndroidDeeplinkDomain(), '')
```
Comment (Giregm) [inline src/features/deeplinks/useDeeplinkUrlHandler.ts:22]: Que penses-tu de `const routeWithParams = url.replace(formatIosDeeplinkDomain(), '').replace(formatAndroidDeeplinkDomain(), '')` 

C'est sans doute peu lisible, je cherchais un moyen d'utiliser des const.

File: src/features/deeplinks/useDeeplinkUrlHandler.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  deeplinkRoutesToScreensMap,
+} from './types'
+
+export function formatIosDeeplinkDomain() {
+  return `${env.URL_PREFIX}://${env.IOS_APP_ID}/`
+}
+
+export function formatAndroidDeeplinkDomain() {
+  return `${env.URL_PREFIX}://${env.ANDROID_APP_ID}/`
+}
```
Comment (Giregm) [inline src/features/deeplinks/useDeeplinkUrlHandler.ts:18]: ici pourquoi ne pas avoir simplement des constantes ?   

```javascript
const ANDROID_PREFIX = `${env.URL_PREFIX}://${env.ANDROID_APP_ID}/`
const IOS_PREFIX = `${env.URL_PREFIX}://${env.IOS_APP_ID}/`  
```

File: src/features/deeplinks/useDeeplinkUrlHandler.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+export function decodeDeeplinkParts(url: string): DeeplinkParts {
+  let route = url.replace(formatIosDeeplinkDomain(), '')
+  route = route.replace(formatAndroidDeeplinkDomain(), '')
+
+  const [routeName, searchParams] = route.split('?') as [AllowedDeeplinkRoutes, string]
```
Comment (Giregm) [inline src/features/deeplinks/useDeeplinkUrlHandler.ts:24]: Ce serait chouette de vérifier que ça match bien une route avant d'y associer le type. Qu'en penses-tu ? 

Cela pourrait faire partie de la condition de validité d'une route

File: src/features/deeplinks/useDeeplinkUrlHandler.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+  const params = searchParams
+    ? JSON.parse(
+        '{"' +
+          decodeURI(searchParams).replace(/"/g, '\\"').replace(/&/g, '","').replace(/=/g, '":"') +
+          '"}'
```
Comment (Giregm) [inline src/features/deeplinks/useDeeplinkUrlHandler.ts:29]: Ici la syntaxe me semble bien se prêter, qu'en penses-tu ?
```javascript
JSON.parse(`{"${ decodeURI(searchParams).replace(/"/g, '\\"').replace(/&/g, '","').replace(/=/g, '":"')}"}"`) 
```

Comment (Giregm) [inline src/features/deeplinks/useDeeplinkUrlHandler.ts:29]: Ici la syntaxe me semble bien se prêter, qu'en penses-tu ?
```javascript
JSON.parse(`{"${ decodeURI(searchParams).replace(/"/g, '\\"').replace(/&/g, '","').replace(/=/g, '":"')}"}"`) 
```

File: src/features/deeplinks/useDeeplinkUrlHandler.test.ts [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+      const deeplinkUrl = formatIosDeeplinkDomain()
+      expect(deeplinkUrl).toEqual(`${env.URL_PREFIX}://${env.IOS_APP_ID}/`)
+    })
+    it('should format properly the deeplink domain for Android', () => {
+      const deeplinkUrl = formatAndroidDeeplinkDomain()
+      expect(deeplinkUrl).toEqual(`${env.URL_PREFIX}://${env.ANDROID_APP_ID}/`)
+    })
+  })
+  describe('Url parts', () => {
+    const uri = 'my-route?param1=one&param2=2&param3=true&param4=false'
```
Comment (Giregm) [inline src/features/deeplinks/useDeeplinkUrlHandler.test.ts:25]: Penses-tu nécessaire de tester certains edge-cases d'uri ? A quel point pourrait-on en rencontrer ?

Je pensais au cas avec un & à la fin par exemple (une erreur sans doute ça reste une uri valide non ? je ne m'y connais pas sur le sujet ainsi je veux bien ton avis !) `'my-route?param1=one&param2=2&param3=true&param4=false&'

File: src/features/deeplinks/useDeeplinkUrlHandler.ts (hunk 3 above)
Comment (Giregm) [reply src/features/deeplinks/useDeeplinkUrlHandler.ts:24]: Je vois que tu le fais dans `useDeeplinkUrlHandler` 

Le seul petit truc qui me "titille" c'est d'assigner le type `AllowedDeeplinkRoutes` à `routeName` sans être sur.

Pour le reste peu importe probablement

File: src/features/deeplinks/useDeeplinkUrlHandler.test.ts [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+  describe('Formatting', () => {
+    it('should format properly the deeplink domain for iOS', () => {
+      const deeplinkUrl = formatIosDeeplinkDomain()
+      expect(deeplinkUrl).toEqual(`${env.URL_PREFIX}://${env.IOS_APP_ID}/`)
+    })
+    it('should format properly the deeplink domain for Android', () => {
+      const deeplinkUrl = formatAndroidDeeplinkDomain()
+      expect(deeplinkUrl).toEqual(`${env.URL_PREFIX}://${env.ANDROID_APP_ID}/`)
+    })
+  })
```
Comment (Giregm) [inline src/features/deeplinks/useDeeplinkUrlHandler.test.ts:23]: Je pense que tu ne testes pas grand chose ici, ces tests ne me paraissent pas nécessaire. A moins que j'ai raté quelque chose !

File: src/features/deeplinks/types.ts [hunk 7]
```diff
@@ -9,9 +9,40 @@ export interface DeeplinkEvent {
...
+> = {
+  [routename in keyof Routes]: {
+    screen: Routes[routename]
+    paramConverter?: (
+      params: Record<string, string>
+    ) => RouteParams<StackParamList, Routes[routename]>
+  }
+}
+
+export const deeplinkToScreenConfiguration: DeepLinksToScreenConfiguration<
```
Comment (Giregm) [inline src/features/deeplinks/types.ts:29]: idem si on a une constante ici, utilise `DEEPLINK_TO_SCREEN_CONFIGURATION` me parait parlant

File: src/features/deeplinks/useDeeplinkUrlHandler.test.ts [hunk 8]
```diff
@@ -74,3 +72,36 @@ describe('useDeeplinkUrlHandler', () => {
...
+        }
+      },
+    },
+    default: {
+      screen: 'Home',
+    },
+  } as DeepLinksToScreenConfiguration<RoutesListForTest, RootStackParamListForTest>,
+}))
+
+/** FAKING NAVIGATION END */
```
Comment (Giregm) [inline src/features/deeplinks/useDeeplinkUrlHandler.test.ts]: Pourquoi ne pas juste mocker `import { RouteParams, RootStackParamList } from 'features/navigation/RootNavigator'` ? De cette manière tu testes aussi le code dans `'features/deeplinks/types'`

File: src/features/home/pages/Home.tsx [hunk 9]
```diff
@@ -26,10 +25,8 @@ import { ACTIVE_OPACITY } from 'ui/theme/colors'
 
 import { SignUpSignInChoiceModal } from '../components/SignUpSignInChoiceModal'
 
-type HomeScreenNavigationProp = StackNavigationProp<RootStackParamList, 'Home'>
-
 type Props = {
-  navigation: HomeScreenNavigationProp
+  navigation: NavigationProp<'Home'>
```
Comment (Giregm) [inline src/features/home/pages/Home.tsx:29]: Tu serais pas chaud pour changer ça en `const {navigation} = useNavigation()` (même si cela sort du scope de ton ticket)

File: src/features/navigation/RootNavigator.tsx [hunk 10]
```diff
@@ -18,8 +19,16 @@ export type RootStackParamList = {
   Home: undefined
   IdCheck: undefined
   Login?: { userId: string }
+  ReinitializePassword: { token: string; expiration_date: number }
 }
 
+export type RouteNames = keyof RootStackParamList
+export type Route<RouteName extends RouteNames> = RouteProp<RootStackParamList, RouteName>
```
[the remark this answers is not in the document]
Comment (blixit) [reply src/features/navigation/RootNavigator.tsx]: je vais utiliser UseRouteType et UseNavigationType pour que ce soit explicite et homogène

File: src/features/deeplinks/types.ts [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+}
+
+export interface DeeplinkEvent {
+  url: string
+}
+
+export type DeepLinkToScreenMap = { [route: string]: RouteNames }
+export const deeplinkRoutesToScreensMap: DeepLinkToScreenMap = {
+  'mot-de-passe-perdu': 'ReinitializePassword',
+}
```
[the remark this answers is not in the document]
Comment (blixit) [reply src/features/deeplinks/types.ts:15]: j'ai hésité à mettre en lettres capitales. je fais ca

[the remark this answers is not in the document]
Comment (blixit) [reply src/features/deeplinks/types.ts:15]: obsolète !

c'est devenu un type entre temps. mais oui le mettre en lettres capitales aurait été meilleur

File: src/features/deeplinks/useDeeplinkUrlHandler.ts (hunk 1 above)
Comment (blixit) [reply src/features/deeplinks/useDeeplinkUrlHandler.ts]: obsolète !

File: src/features/deeplinks/useDeeplinkUrlHandler.ts (hunk 2 above)
Comment (blixit) [reply src/features/deeplinks/useDeeplinkUrlHandler.ts]: obsolète !

File: src/features/deeplinks/useDeeplinkUrlHandler.ts (hunk 3 above)
Comment (blixit) [reply src/features/deeplinks/useDeeplinkUrlHandler.ts:14]: je comprends ton point.

cependant, TypeScript a du mal avec l'expression `DEEPLINK_TO_SCREEN_CONFIGURATION[routeName]` si routeName est une string.

par contre, tu verras que je fais la vérification de l'existence du screen avant de continuer.

chaud pour qu'on en parle de vive voix

File: src/features/deeplinks/useDeeplinkUrlHandler.test.ts (hunk 5 above)
Comment (blixit) [reply src/features/deeplinks/useDeeplinkUrlHandler.test.ts:10]: j'y ai pensé ! par forcément au & à la fin mais à implémenter une fonction plus rigoureuse mais en réalité ca n'apporte aucune plus value.

nos urls sont connues et ont une syntaxe validée par le back. Je pense que si on se retrouve avec une url fausse d'un caractère, ca serait plutôt aux paramétreurs de la corriger

File: src/features/deeplinks/useDeeplinkUrlHandler.ts (hunk 4 above)
Comment (blixit) [reply src/features/deeplinks/useDeeplinkUrlHandler.ts:19]: bien vu

Comment (blixit) [reply src/features/deeplinks/useDeeplinkUrlHandler.ts:19]: doublon

File: src/features/deeplinks/useDeeplinkUrlHandler.test.ts (hunk 6 above)
Comment (blixit) [reply src/features/deeplinks/useDeeplinkUrlHandler.test.ts]: obsolète ! code déplacé !

je pense que c'est pertinent dans la mesure où on se protège contre une évolution ne serait-ce qu'accidentelle de cette concaténation (une modification pendant un test et qui passe la revue, ... ce genre de chose).

File: src/features/deeplinks/useDeeplinkUrlHandler.test.ts (hunk 8 above)
Comment (blixit) [reply src/features/deeplinks/useDeeplinkUrlHandler.test.ts]: @Giregm enfin on ne peut pas mocker un type. enfin, j'ai pas trouver comment faire !

>De cette manière tu testes aussi le code dans 'features/deeplinks/types'
bah c'est déjà le cas. je teste que la mécanique de la configuration fonctionne bien. Tester les routes utilisées dans la vraie configuration reviendrait seulement à faire une sorte de snapshot, ce qui n'a pas de plus value ici à mon avis

File: src/features/cheatcodes/pages/Navigation.tsx [hunk 12]
```diff
@@ -18,7 +23,32 @@ export function Navigation(): JSX.Element {
...
+        />
+        <LittleDescription>{_(t`simulate email deeplink clic with:`)}</LittleDescription>
+        <LittleDescription>{MdpDeeplink}</LittleDescription>
+        <Spacer.Column numberOfSpaces={5} />
       </ScrollView>
     </SafeContainer>
   )
 }
+
+export default Navigation
```
Comment (Giregm) [inline src/features/cheatcodes/pages/Navigation.tsx:43]: à enlever, on exporte déjà la fonction `Navigation`

File: src/features/deeplinks/useDeeplinkUrlHandler.ts (hunk 3 above)
Comment (Giregm) [reply src/features/deeplinks/useDeeplinkUrlHandler.ts:14]: vu ensemble, typeguard pour checker que string est AllowedDeep..

File: src/features/deeplinks/utils.test.ts [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+  it('should format properly the deeplink domain for iOS', () => {
+    Platform.OS = 'ios'
+    const deeplinkUrl = formatDeeplinkDomain()
+    expect(deeplinkUrl).toEqual(`${env.URL_PREFIX}://${env.IOS_APP_ID}/`)
+  })
+  it('should format properly the deeplink domain for Android', () => {
+    Platform.OS = 'android'
+    const deeplinkUrl = formatDeeplinkDomain()
+    expect(deeplinkUrl).toEqual(`${env.URL_PREFIX}://${env.ANDROID_APP_ID}/`)
+  })
```
Comment (Giregm) [inline src/features/deeplinks/utils.test.ts:11]: vu ensemble, test ko

File: src/ui/components/BottomCard.tsx [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+    onBeforeHide(data) {
+      if (!data.keyboardShown) {
+        setKeyboardHeight(0)
+      }
+    },
+  })
+
+  return (
+    <TouchableWithoutFeedback onPress={Keyboard.dismiss}>
+      <AvoidingKeyboardContainer bottom={keyboardHeight}>
```
Comment (Giregm) [inline src/ui/components/BottomCard.tsx:26]: Que penses-tu d'appeler la prop `bottom` quelque chose comme `keyboardHeight`?

File: src/ui/components/BottomCard.tsx [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+interface AvoidingKeyboardContainerProps {
+  bottom: number
+}
+const AvoidingKeyboardContainer = styled(View)<AvoidingKeyboardContainerProps>(({ bottom }) => ({
+  bottom,
+}))
+
+const StyledFakeModal = styled.View({
```
Comment (Giregm) [inline src/ui/components/BottomCard.tsx:41]: naming, `StyledBottomCardContainer` ?

File: src/ui/components/keyboard/useKeyboardEvents.ts [hunk 16]
```diff
@@ -0,0 +1,72 @@
+// based on https://github.com/react-native-community/hooks/blob/master/src/useKeyboard.ts
+import { useEffect } from 'react'
+import { Keyboard, KeyboardEventListener, ScreenRect, Platform } from 'react-native'
+
+interface UseKeyboardReturnData {
```
Comment (Giregm) [inline src/ui/components/keyboard/useKeyboardEvents.ts:5]: vu ensemble, il faut ajouter quelques tests

File: src/features/auth/pages/ReinitializePassword.tsx [hunk 17]
```diff
@@ -35,50 +36,56 @@ export const ReinitializePassword = () => {
   }
 
   useEffect(() => {
-    if (expiration_date < new Date().getTime()) {
+    if (expiration_date * 1000 < new Date().getTime()) {
```
Comment (Giregm) [inline src/features/auth/pages/ReinitializePassword.tsx:39]: `*1000` d'ou vient ce nombre ? un nom de constante peut servir
`const MILLISECONDS_IN_A_SECOND`

File: src/features/deeplinks/utils.ts [hunk 18]
```diff
@@ -1,7 +1,5 @@
-import { Platform } from 'react-native'
-
 import { env } from 'libs/environment'
 
 export function formatDeeplinkDomain() {
-  return `${env.URL_PREFIX}://${Platform.OS === 'ios' ? env.IOS_APP_ID : env.ANDROID_APP_ID}/`
+  return `${env.URL_PREFIX}://app.${env.URL_PREFIX}.${env.ENV}/`
```
Comment (Giregm) [inline src/features/deeplinks/utils.ts:4]: Je suis pour utiliser une constante, quit à en faire une snapshot/ un test 
+ ajouter un commentaire sur le test qui fait le lien avec la nécessité de ne changer la constante qu'en accord avec le backend

Comment (blixit) [reply src/features/deeplinks/utils.ts:4]: j'ai remplacé par une constante

File: src/features/deeplinks/typeGuard.ts [hunk 19]
```diff
@@ -0,0 +1,5 @@
+import { AllowedDeeplinkRoutes, DEEPLINK_TO_SCREEN_CONFIGURATION } from './types'
+
+export const isAllowedRouteTypeguard = (routeName: string): routeName is AllowedDeeplinkRoutes => {
```
Comment (Giregm) [inline src/features/deeplinks/typeGuard.ts:3]:
```suggestion
export const isAllowedRouteTypeGuard = (routeName: string): routeName is AllowedDeeplinkRoutes => {
```
Si on veut être (preque) trop rigoureux ^^

File: src/features/deeplinks/typeGuard.ts [hunk 20]
```diff
@@ -0,0 +1,5 @@
...
+  return (typeof routeName === 'string' && routeName in DEEPLINK_TO_SCREEN_CONFIGURATION) ?? false
```
Comment (Giregm) [inline src/features/deeplinks/typeGuard.ts:4]: Il me semble que ` (typeof routeName === 'string' && routeName in DEEPLINK_TO_SCREEN_CONFIGURATION)` est déjà un booléen. Ainsi je n'aurais pas mis `?? false` 

Qu'en penses-tu ?

File: src/features/deeplinks/useDeeplinkUrlHandler.test.ts (hunk 8 above)
Comment (Giregm) [reply src/features/deeplinks/useDeeplinkUrlHandler.test.ts]: okpm :D
